timemachine/cpp/src/kernels/k_exchange.cu

```diff
@@ -0,0 +1,689 @@
+        } else {
+            targeting_inner_volume[0] = 0;
+        }
+    } else {
+        assert(0);
+    }
+}
+
+template void __global__ k_decide_targeted_move<float>(
+    const int num_target_mols,
+    const float *__restrict__ rand,
+    const int *__restrict__ inner_count,
+    int *__restrict__ targeting_inner_volume);
+template void __global__ k_decide_targeted_move<double>(
+    const int num_target_mols,
+    const double *__restrict__ rand,
+    const int *__restrict__ inner_count,
+    int *__restrict__ targeting_inner_volume);
+
+// k_separate_weights_for_targeted takes the flag and the mol indices and writes them out
+// to a new buffer the weights associated with the source molecules.
+template <typename RealType>
+void __global__ k_separate_weights_for_targeted(
+    const int num_target_mols,
+    const int *__restrict__ targeting_inner_volume, // [1]
+    const int *__restrict__ inner_count,            // [1]
+    const int *__restrict__ partitioned_indices,    // [inner_count]
+    const RealType *__restrict__ weights,           // [num_target_mols]
+    RealType *__restrict__ output_weights) {
+    int idx = blockIdx.x * blockDim.x + threadIdx.x;
+    const int target_inner = targeting_inner_volume[0];
+    const int local_inner_count = inner_count[0];
+    const int outer_count = num_target_mols - local_inner_count;
+    const int count = target_inner == 1 ? outer_count : local_inner_count;
+    const int offset = target_inner == 1 ? local_inner_count : 0;
+    while (idx < count) {
+        output_weights[idx] = weights[partitioned_indices[idx + offset]];
+
+        idx += gridDim.x * blockDim.x;
+    }
+}
+
+template void __global__ k_separate_weights_for_targeted<float>(
+    const int num_target_mols,
+    const int *__restrict__ targeting_inner_volume, // [1]
+    const int *__restrict__ inner_count,            // [1]
+    const int *__restrict__ partitioned_indices,    // [inner_count]
+    const float *__restrict__ weights,              // [num_target_mols]
+    float *__restrict__ output_weights);
+template void __global__ k_separate_weights_for_targeted<double>(
+    const int num_target_mols,
+    const int *__restrict__ targeting_inner_volume, // [1]
+    const int *__restrict__ inner_count,            // [1]
+    const int *__restrict__ partitioned_indices,    // [inner_count]
+    const double *__restrict__ weights,             // [num_target_mols]
+    double *__restrict__ output_weights);
+
+template <typename RealType>
+void __global__ k_setup_destination_weights_for_targeted(
+    const int num_target_mols,
+    const int *__restrict__ samples,                // [1]
+    const int *__restrict__ targeting_inner_volume, // [1]
+    const int *__restrict__ inner_count,            // [1]
+    const int *__restrict__ partitioned_indices,    // [inner_count]
+    const RealType *__restrict__ weights,           // [num_target_mols]
+    RealType *__restrict__ output_weights) {
```

Comment:
document size of `output_weights` as `[num_target_mols]` but maybe note that we only use up to size `count + 1`

Reply:
Added in https://github.com/proteneer/timemachine/pull/1205/commits/6c7f11565140ec23d0bbda7731e9a63d4ba0b041